### PR TITLE
feat(show): diff delta requirements against the main specs

### LINKED ISSUE #427: Feature Request: Provide an `openspec diff` command to show differences between a spec and its modified version.
Provide an openspec diff command or a similar feature to show differences between a spec and its modified version.

Currently, we can manually copy `/openspec/changes/*/specs/A.md` and paste it into `/openspec/specs/A/spec.md` to compare it with the existing spec.

It would be very useful to have a built-in diff feature to automatically highlight these differences.

Thanks for this powerful tool! ❤️

### PR DESCRIPTION
**Pull request by bsmedberg-xometry**
## Status

LGTM for final review. Functional hardening, the combined pnpm lockfile, and the CI-derived Nix dependency hash all pass every CI/security/platform gate.

## What was missing

A MODIFIED requirement must restate its complete block, so reviewers had no direct view of what changed relative to the main spec. The earlier implementation also had unsafe edge cases: JSON could lose the diff when a warning existed, discovery/read failures could be treated as missing data, chained renames could lose the original requirement name, and an empty diff was indistinguishable from no diff.

## What it does

`openspec show <change> --diff` renders each delta requirement against the selected root's main spec:

- MODIFIED requirements receive a unified diff with range headers that keep distant hunks unambiguous, including explicit empty diffs.
- ADDED, REMOVED, and RENAMED requirements retain their authored meaning.
- JSON MODIFIED entries can carry both `diff` and `warning`.
- Chained renames resolve back to the original main-spec name.
- Exact matches win; folded-name fallbacks warn because archive matches exactly.
- Missing main specs/requirements warn, while discovery and non-ENOENT read failures fail closed.
- Nested capabilities and registered stores resolve through the same root-selection path.
- Without `--diff`, existing output remains unchanged.

This branch includes #1718's development-dependency update and regenerated pnpm lockfile. Merge #1718 before this PR; the stacked dependency commit then drops out cleanly.

## Proof it works

- Diff/show regressions: 36/36 passed, including distant multi-hunk output.
- Broader show/parser/diff/store suites: 120/120 passed.
- Version checks: 48/48 passed outside the network-restricted sandbox.
- Build, lint, strict OpenSpec validation, and frozen pnpm 9.15.9 install pass.
- Production and full dependency audits report no known vulnerabilities.
- Full local suite reached 4,093 passing tests; remaining failures reproduce on clean main or require listeners denied by the sandbox.
- Linux, macOS, Windows, security, lint, release tracking, and CodeRabbit checks pass.
- Nix validates the combined dependency hash `sha256-+qGFLSVLJ9faZOmfO6ZVBP525i5LRgwhsJat2vT7Aw8=`.

## Notes / nits

Adds `diff` (jsdiff v9, MIT) as a runtime dependency. Diffs compare files in the selected OpenSpec root, not git history, and cover spec deltas only.

Closes #427

### CODE REVIEW

File: openspec/changes/spec-diffs/specs/cli-show/spec.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+- **WHEN** executing `openspec show <change-name>` in text mode (no `--json`)
+- **THEN** display the proposal markdown text
+- **AND** if delta spec files exist under `openspec/changes/<change-name>/specs/`, display each delta spec's full content grouped by capability name
+
+#### Scenario: Diff output in text mode
+
+- **WHEN** executing `openspec show <change-name> --diff` in text mode (no `--json`)
+- **THEN** display the proposal markdown text
+- **AND** for each delta spec file under `openspec/changes/<change-name>/specs/<cap>/spec.md`, iterate parsed deltas grouped by capability
```
[the remark this answers is not in the document]
Comment (bsmedberg-xometry) [reply openspec/changes/spec-diffs/specs/cli-show/spec.md:46]: Actually, the spec was correct: I want `openspec show --diff` to print the proposal and then the diffs of the changed requirements... it will display new requirements completely.

File: src/commands/change.ts [hunk 2]
```diff
@@ -71,21 +92,262 @@ export class ChangeCommand {
...
+
+      const plan = parseDeltaSpec(deltaContent);
+      const renameMap = buildRenameMap(plan.renamed);
+
+      for (const block of plan.added) {
+        results.push({ capability: capName, operation: 'ADDED', requirementName: block.name, raw: block.raw });
+      }
+
+      for (const name of plan.removed) {
+        results.push({ capability: capName, operation: 'REMOVED', requirementName: name, raw: `### Requirement: ${name}` });
```
[the remark this answers is not in the document]
Comment (bsmedberg-xometry) [reply src/commands/change.ts:268]: My intention was to keep displaying removed requirements as they are today: just with the name, not with the full body. So the spec and the implementation match, but I'm open to human maintainer feedback that it might be useful to display the full body of the removed requirement.

File: test/commands/show-diff.test.ts [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+        encoding: 'utf-8',
+        env: { ...process.env, NO_COLOR: '1' },
+        stdio: ['pipe', 'pipe', 'pipe'],
+      });
+      return { stdout, stderr: '' };
+    } catch (e: any) {
+      return { stdout: e.stdout?.toString() ?? '', stderr: e.stderr?.toString() ?? '' };
+    } finally {
+      process.chdir(originalCwd);
+    }
```
[the remark this answers is not in the document]
Comment (bsmedberg-xometry) [reply test/commands/show-diff.test.ts:127]: This is a common pattern across the test suite (used in spec.test.ts, show.test.ts, and our show-diff.test.ts).  Since this is not new to the tests in this PR, I'm going to decline this recommendation for now.

**Review by alfred-openspec — DISMISSED** [review]
Thanks for pushing this. The reviewer pain is real and `show --diff` is the right shape, but I don't think this is mergeable yet because the implementation still violates the PR's own output contract in a few places:

- Text mode `--diff` still prints `proposal.md` before the diff because `ChangeCommand.show()` reads/logs the proposal before branching into `showSpecDiffs()`. The PR says `--diff` without `--json` should be diff-only.
- `showSpecDiffs()` silently returns when there are no delta specs. The spec says this should emit a successful user-facing message.
- REMOVED requirements lose their removal body. The result is synthesized as `### Requirement: ${name}`, so Reason/Migration text is dropped from diff output.
- `diffRequirementBlock()` still calls `.trim()`, which can strip leading whitespace from the first context line in a hunk. This should be `trimEnd()` so unified diff content stays faithful.
- The new tests currently lock in the wrong text-mode behavior by asserting proposal text appears for `--diff`, and the shell-based helpers should avoid `execSync` string interpolation for path safety.

I'd fix those before merge. Once the CLI contract, JSON/text edge cases, and tests line up, I think this becomes a solid incremental feature without changing the storage model.

**Comment by clay-good** [comment]
Picked this up and pushed a merge with `main` plus fixes for the review points. Point by point:

1. **Text mode `--diff` prints the proposal first** — kept, and the contract now says so in one place. `proposal.md` + diffs is what `openspec/changes/spec-diffs/proposal.md` specified; the "diffs only" line was a leftover in the PR description and is gone. The real behavior problem was the other branch: the previous revision printed *full delta spec content* on plain `openspec show <change>`, changing existing output for everyone. That is reverted — without `--diff` the command is a raw proposal passthrough again, pinned by a test that compares stdout to `proposal.md` byte for byte.
2. **Silent return with no delta specs** — now prints `No delta specs to diff for change "<name>".` and exits 0, with a test.
3. **REMOVED loses Reason/Migration** — fixed at the parser: `parseDeltaSpec` returns `removedBlocks` alongside `removed`, so the authored block (Reason, Migration) is printed under the red label. Bullet-list removals have no body and stay name-only.
4. **`.trim()` in `diffRequirementBlock`** — already `trimEnd()` as of 3fb6795; there is a test asserting the leading space survives on the first context line.
5. **Tests locking in wrong behavior / `execSync` interpolation** — the text-mode expectations follow the corrected contract, and every CLI call now goes through `execFileSync`/`spawnSync` with an argv array from a `mkdtemp` project instead of an interpolated string and an in-repo temp directory.

Two things beyond the list, both hardening:

- **Store awareness.** Main specs were resolved from `process.cwd()`, so `openspec show <change> --store <id> --diff` diffed against whatever `openspec/specs` happened to sit in the working directory. They now resolve through the command's root. The regression test plants a decoy `billing` spec in the cwd repo and asserts the diff comes from the store.
- **One collection pass.** Text mode and JSON mode had two near-identical implementations of the same lookup. They now share `collectSpecDiffs()`, so the two surfaces cannot drift; a new test with two MODIFIED requirements in one spec pins that each diff lands on its own delta.

`nix build` and `nix run . -- --version` verified locally with the regenerated `pnpmDeps` hash. Suite: 4,101 pass; the two failures (`artifact-workflow` Cursor skills, `config-profile` in-process apply) reproduce on a clean `origin/main` checkout.

**Comment by clay-good** [comment]
One more fix on top of the review round: `collectSpecDiffs()` enumerated only top-level directories under the change's `specs/`, so a nested capability (`specs/<area>/<id>/spec.md`, supported since #1353) was skipped entirely — no text output, and its MODIFIED deltas came back from `--json` with no `diff` field. It now uses the same `discoverSpecFiles()` helper `ChangeParser` uses, which also keeps the capability ids identical to the `spec` field of the JSON deltas. The regression test fails on the old enumeration and passes on the new one. Suite: 4,102 pass, same two pre-existing `main` failures.

**Comment by clay-good** [comment]
Working through CodeRabbit's latest pass — all three land, two as code and one as a contract correction:

**Missing main spec on a MODIFIED requirement** (`change.ts`) — was rendered as all-additions, which reads like a new capability. A MODIFIED requirement names a block that is supposed to already exist, so that combination is an authoring error archive will reject. It now prints the raw text with `No main spec at openspec/specs/<cap>/spec.md, so MODIFIED requirement "<name>" has nothing to diff against`, and JSON gets `warning` with no `diff`.

**Interior whitespace in requirement names** — real, and the fix runs the other direction from "just normalize it". `openspec` matches requirement names *exactly* (`normalizeRequirementName` trims and nothing else; `foldRequirementName` exists precisely because a case- or spacing-only difference "means a mistake, never two requirements"). Loosening the diff lookup would have made `--diff` succeed where archive fails, hiding the problem. The lookup is now exact first, then the shared fold as a fallback, and a folded match prints **both** the diff the author meant and a warning: `Header differs from the main spec's "User login" only in case or spacing; archive matches names exactly, so reconcile them before archiving`. The previous revision's `.toLowerCase()` comparison had the same blind spot and is gone.

**REMOVED contract in `proposal.md`** — corrected; the text said "shows the requirement name" while `printDiffText()` prints the authored removal block. The proposal now describes the rendered block, Reason and Migration included.

Nothing outstanding from your earlier round: the JSON contract in the proposal already spells out MODIFIED-only `diff` and the `warning` fallback, and the text-mode "proposal first" question was settled upthread (proposal first, diffs after) — the actual defect there was the old revision printing delta content on plain `openspec show <change>`, which is reverted and pinned by a test.

Suite: 4,106 pass; the two remaining failures reproduce on a clean `origin/main` worktree and pass in CI.

**Review by clay-good — DISMISSED** [review]

**Review by clay-good — APPROVED** [review]
